pkg/ingester/ingester.go

```diff
@@ -1035,31 +1065,46 @@ func (i *Ingester) startPushRequest(ctx context.Context, reqSize int64) (_ conte
 	return ctx, true, nil
 }
 
-func (i *Ingester) finishPushRequest(reqSize int64) {
+func (i *Ingester) finishPushRequest(ctx context.Context, reqSize int64, duration time.Duration, err error) {
 	i.inflightPushRequests.Dec()
 	if reqSize > 0 {
 		i.inflightPushRequestsBytes.Sub(reqSize)
 	}
+	_ = i.circuitBreaker.finishPushRequest(ctx, duration, err)
 }
 
 // PushWithCleanup is the Push() implementation for blocks storage and takes a WriteRequest and adds it to the TSDB head.
-func (i *Ingester) PushWithCleanup(ctx context.Context, req *mimirpb.WriteRequest, cleanUp func()) error {
+func (i *Ingester) PushWithCleanup(ctx context.Context, req *mimirpb.WriteRequest, cleanUp func()) (err error) {
 	// NOTE: because we use `unsafe` in deserialisation, we must not
 	// retain anything from `req` past the exit from this function.
 	defer cleanUp()
 
+	start := time.Now()
 	// Only start/finish request here when the request comes NOT from grpc handlers (i.e., from ingest.Store).
 	// NOTE: request coming from grpc handler may end up calling start multiple times during its lifetime (e.g., when migrating to ingest storage).
 	// startPushRequest handles this.
 	if i.cfg.IngestStorageConfig.Enabled || !i.cfg.LimitInflightRequestsUsingGrpcMethodLimiter {
 		reqSize := int64(req.Size())
-		_, shouldFinish, err := i.startPushRequest(ctx, reqSize)
-		if err != nil {
-			return middleware.DoNotLogError{Err: err}
+		_, shouldFinish, startPushErr := i.startPushRequest(ctx, reqSize)
+		if startPushErr != nil {
+			return middleware.DoNotLogError{Err: startPushErr}
 		}
 		if shouldFinish {
-			defer i.finishPushRequest(reqSize)
+			defer func() {
+				i.finishPushRequest(ctx, reqSize, time.Since(start), err)
+			}()
 		}
+	} else {
+		defer func() {
+			// We enrich the pushRequestState contained in the context with this PushWithCleanUp()
+			// call duration, and a possible error it returns. These data are needed during a
+			// successive call to FinishPushRequest().
+			if st, ok := ctx.Value(pushReqCtxKey).(*pushRequestState); ok {
+				st.requestDuration = time.Since(start)
+				st.err = err
+				ctx = context.WithValue(ctx, pushReqCtxKey, st)
```

Comment:
We don't need to create new context here. We're modifying value that's already in the context. This new context is not used anywhere.